Framework/Algorithms/src/RemoveInstrumentGeometry.cpp

```diff
@@ -0,0 +1,108 @@
+using Mantid::Kernel::Direction;
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(RemoveInstrumentGeometry)
+
+//----------------------------------------------------------------------------------------------
+
+/// Algorithms name for identification. @see Algorithm::name
+const std::string RemoveInstrumentGeometry::name() const { return "RemoveInstrumentGeometry"; }
+
+/// Algorithm's version for identification. @see Algorithm::version
+int RemoveInstrumentGeometry::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string RemoveInstrumentGeometry::category() const { return "Utility\\Workspaces"; }
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string RemoveInstrumentGeometry::summary() const {
+  return "Removes instrument geometry records from a given workspace.";
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void RemoveInstrumentGeometry::init() {
+  declareProperty(std::make_unique<WorkspaceProperty<API::Workspace>>("InputWorkspace", "", Direction::Input),
+                  "An input workspace.");
+  declareProperty(std::make_unique<WorkspaceProperty<API::Workspace>>("OutputWorkspace", "", Direction::Output),
+                  "An output workspace.");
+  declareProperty(std::make_unique<Kernel::ArrayProperty<int>>("MDExperimentInfoNumbers"),
+                  "For MD workspaces, the ExperimentInfo indices to have the instrument removed."
+                  "If empty, the instrument will be removed from all ExperimentInfo objects."
+                  "The parameter is ignored for any other workspace type.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void RemoveInstrumentGeometry::exec() {
+  Workspace_const_sptr inputWS = this->getProperty("InputWorkspace");
+
+  Workspace_sptr outputWS = this->getProperty("OutputWorkspace");
+  if (outputWS != inputWS) {
+    outputWS = inputWS->clone();
+  }
+
+  // create an empty instrument
+  Instrument_sptr emptyInstrument(new Geometry::Instrument());
+
+  MatrixWorkspace_sptr outputMtrxWS = std::dynamic_pointer_cast<API::MatrixWorkspace>(outputWS);
+  if (outputMtrxWS != nullptr) { // it is a matrix workspace
+    outputMtrxWS->setInstrument(emptyInstrument);
+  } else {
+    MultipleExperimentInfos_sptr outputMDWS = std::dynamic_pointer_cast<Mantid::API::MultipleExperimentInfos>(outputWS);
+    if (outputMDWS != nullptr) { // it is an MD workspace
+                                 // for which experiment indices do we need to remove the instrument geometry?
+      std::vector<uint16_t> indicesToRemoveInstrument;
+      std::string indicesToRemoveInstrument_str = this->getProperty("MDExperimentInfoNumbers");
+      if (!indicesToRemoveInstrument_str.empty()) {
```

Comment:
You don't need to parse the string. The `ArrayProperty<int>` can do it for you. Just call the `value` method, and it will return an `std::vector<int>`

Reply:
I have simplified the algorithm and updated the branch.